src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netcore.cs

```diff
@@ -7760,10 +9220,18 @@ public static Vector512<float> Invoke(Vector512<float> x, Vector512<float> y)
             public static Vector512<float> Invoke(Vector512<float> x, Vector512<float> y) => x * y;
 #endif
 
-            public static float Invoke(Vector128<float> x) => HorizontalAggregate<MultiplyOperator>(x);
-            public static float Invoke(Vector256<float> x) => HorizontalAggregate<MultiplyOperator>(x);
+            public static float Invoke(Vector128<float> x)
+            {
+                // We need to do log2(count) operations to compute the total sum
+
+                x *= Vector128.Shuffle(x, Vector128.Create(2, 3, 0, 1));
+                x *= Vector128.Shuffle(x, Vector128.Create(1, 0, 3, 2));
```

Comment:
This ends up doing two shuffles and two multiplications instead of three multiplications. I assume that tradeoff is a good one, especially if with the individual element loads to do the three multiplications?

Reply:
Right. And noting it's not just 3 multiplications but rather 3 extracts + 3 multiplications.

x64 doesn't have a "proper" floating-point extract operation. It has `extractps` but that moves the value into an integer register, so it doesn't get used. Instead, `x.GetElement(n)` works as a `nop` (0), `movshdup` (1), `unpckhps` (2), or `shufps` (3). All of these are 1 cycle operations, with 1 reciprocal throughput, and which execute on port 5. `mulps` is then a 4 cycle operation with a 0.5 reciprocal throughput, and which can execute on port 0 or 1.

So we're effectively trading off `(3 * 1) shuffle cycles + (3 * 4) multiply cycles = 15 cycles` for `(2 * 1) shuffle cycles + (2 * 4) multiply cycles = 10 cycles`